add: schema v2 documentation (#6)
ilikesymmetry/ERCs: ERCS/erc-8021.md

Custom Code Registries can be referenced across chains for ease of connecting identity for multi-chain apps.

> Schema ID `2` uses CBOR encoding for extensible transaction annotation:

Reviewer: suggestion: Have [CBOR](https://cbor.io/) link to its website for more info

1. **cborLength** (2 bytes): Length of the CBOR-encoded map in bytes (big-endian, always 2 bytes regardless of value)
2. **cborData** (variable length): CBOR-encoded map containing attribution data

> The CBOR map MUST use the following key schema. All top-level keys are reserved for this specification.

Reviewer 2: Can we make explicit that all fields are optional. On read it feels like all are required

| Key | Description                                                                                                                             |
| --- | --------------------------------------------------------------------------------------------------------------------------------------- |

> | `a` | Builder code (application) - string value                                                                                               |

Reviewer: issue: This is the first time in the doc that we use the term builder code. The rest of the doc refers to them as attribution codes or simply codes. IMO, we should be consistent in terms here. I would personally go with app code, wallet code, etc, but not a strong opinion

Reviewer 2: +1 to no mention of branded product name "Builder Codes". Would prefer "App code" to make a mental link with the key `a` as well.

> | `a` | Builder code (application) - string value                                                                                               |

Reviewer: suggestion: Call out where fields are required or optional. I would also consider putting type on its own column

Author: This is difficult cause technically everything is optional. You could have a wallet code with no app code ... i guess custom registry is optional and metadata... but nothing is required

Reviewer: IMO, wallet code with no app code would be totally valid. The only wonkyness would be if someone sent a custom registry for app/wallet but no app/wallet code, but the parser could just ignore that.

> | `w` | Wallet builder code - string value                                                                                                      |

Reviewer 2: ```suggestion
| `w` | Wallet code - string value                                                                                                      |
```

> | `r` | Custom registries - sub-map keyed by entity type (`a` for application, `w` for wallet), each containing `c` (chainId) and `a` (address) |

Reviewer: issue: We should be explicit that address is supposed to be a 20-byte byte string somewhere.

| `m` | Custom metadata - sub-map of arbitrary key-value pairs                                                                                  |

The `r` (registry) field structure:

```json
{
  "r": {
    "a": { "c": <chainId>, "a": "<address>" },
    "w": { "c": <chainId>, "a": "<address>" }
  }
}

> ```

Reviewer 2: Think we should show full ts type for the entire table enumerated above which can include this:
```ts
type Schema2Data = {
  a?: string
  w?: string
  r?: {
    a?: {c: Hex, a: Hex}
    w?: {c: Hex, a: Hex}
  }
  m?: Record<any, any>
}
```

> Only the `m` (metadata) field MAY contain arbitrary data defined by consumers of this specification. All other top-level keys are reserved and MUST NOT be used for purposes other than those defined above.

Reviewer 2: Feels a bit weird to do an "only...MAY" combo, how about:

```suggestion
Any other data included in the schema data MUST be included within the `m` (metadata) field. All other top-level keys are reserved and MUST NOT be used for purposes other than those defined above or in future iterations of the standard.
```

Schema ID `2` enables:

- Optional and unknown fields for future extensions
- Coexistence with other suffix-based systems

> - Arbitrary metadata without requiring new schema IDs

Reviewer 2: feels weird to put the motivation at the end, consider putting at top of its section

Note: Schema ID `2` trades minimal byte size for extensibility. Applications prioritizing minimal calldata overhead SHOULD use Schema ID `0` or `1` when their use case does not require extensibility.

#### Examples

The complete suffix format is:

```
{txData}{schemaData}{schemaId}{ercSuffix}
```

Single entity attribution with canonical registry:

```
{txData}{"baseapp"}{7}{0}{0x80218021802180218021802180218021}
```

Multiple entity attribution with custom registry:

```
{txData}{0xCodeRegistry}{8453}{2}{"baseapp,morpho"}{14}{1}{0x80218021802180218021802180218021}
```

Single entity attribution with CBOR (Schema ID 2):

```

> {txData}{cborData}{0x000c}{2}{0x80218021802180218021802180218021}

Reviewer 2: Cut the map below and put in the example for obviousness
```suggestion
{txData}{{"a": "baseapp"}}{0x000c}{2}{0x80218021802180218021802180218021}
```

```

CBOR map: `{"a": "baseapp"}`

Multiple entities with wallet, custom registry, and metadata (Schema ID 2):

```
{txData}{cborData}{0x0085}{2}{0x80218021802180218021802180218021}
```

CBOR map:

```json

> {

Reviewer 2: Think we should do something like this where it's all one connected code block for readability:
```
{txData}{cborData}{0x0085}{2}{0x80218021802180218021802180218021}

// cborData
{
  "a": ...
   ...
}
```

  "a": "baseapp",
  "w": "privy",
  "r": {
    "a": {

>       "c": 8453,

Reviewer 2: Precedent for `chainId` is generally hex. Is there a type for bytes in CBOR or do we need to do ints?

Only the `m` (metadata) field MAY contain arbitrary data defined by consumers of this specification. All other top-level keys are reserved and MUST NOT be used for purposes other than those defined above.

```typescript
type Schema2Data = {
  a?: string;
  w?: string;
  r?: {
    a?: { c: Hex; a: Hex };
    w?: { c: Hex; a: Hex };
  };
  m?: Record<string, unknown>;
};
```

Note: Schema ID `2` trades minimal byte size for extensibility. Applications prioritizing minimal calldata overhead SHOULD use Schema ID `0` or `1` when their use case does not require extensibility.

#### Examples

The complete suffix format is:

```
{txData}{schemaData}{schemaId}{ercSuffix}
```

Single entity attribution with canonical registry:

```
{txData}{"baseapp"}{7}{0}{0x80218021802180218021802180218021}
```

Multiple entity attribution with custom registry:

```
{txData}{0xCodeRegistry}{8453}{2}{"baseapp,morpho"}{14}{1}{0x80218021802180218021802180218021}
```

Single entity attribution with CBOR (Schema ID 2):

```
{txData}{"a":"baseapp"}{0x000b}{2}{0x80218021802180218021802180218021}
```

Multiple entities with wallet, custom registry, and metadata (Schema ID 2):

```

> {txData}{"a":"baseapp","w":"privy","r":{"a":{"c":"0x2105","a":"0xBcf2B9598Ec0781eE49230CaACF0FB3C881B5195"}},"m":{"utm_campaign":"winter-promo","source":"webapp"}}{0x007b}{2}{0x80218021802180218021802180218021}

Reviewer 2: This is a disaster to read lol. Maybe we should use `|` to separate sections so it's at least a different symbol. What do you think of this style? If it feels good, would need to update the other examples to be consistent.

```suggestion
txData|{"a":"baseapp","w":"privy","r":{"a":{"c":"0x2105","a":"0xBcf2B9598Ec0781eE49230CaACF0FB3C881B5195"}},"m":{"utm_campaign":"winter-promo","source":"webapp"}|0x007b|2|0x80218021802180218021802180218021
```

Author: lol i think that or where you break out the object...

Objects are easier to read when properly formatted in my opinion

Reviewer 2: move the object out to read separately, but keep in same code block so it's easier to read them as one grouping?

Author: Yeah basically we could have the old format with {CborData} and then in that code block show the structured object

   e. Extract the previous 20 bytes as `codeRegistryAddress`
   f. Parse `codes` using `0x2C` (comma) as delimiter between codes
   g. Query code-related data from `codeRegistryAddress` on chain with id `codeRegistryChainId`
5. If `schemaId` is `2`:

>    a. Extract the previous 2 bytes as `cborLength` (big-endian)

Reviewer: super-nit: First time mentioning endianness, should either update it everywhere for consistency or remove it. I would probably just include it once at the top that all codesLength (inception suggestion: standardize on the terminology for codesLength/cborLength) use big-endian order

   b. Extract the previous `cborLength` bytes as `cborData`
   c. Decode `cborData` as a CBOR map

>    d. Extract builder code from key `a`

Reviewer 2: ```suggestion
   d. Extract app code from key `a`
```

- codeRegistryAddress: 0xcccccccccccccccccccccccccccccccccccccccc
```

Single entity attribution with CBOR (Schema ID 2):

```
Input: 0xdddddddd{cborData}000c0280218021802180218021802180218021
Expected:
- txData: 0xdddddddd
- schemaId: 2
- cborLength: 0x000c (12 bytes, 2-byte big-endian)
- cborData: {"a": "baseapp"}
- builderCode: "baseapp"

> - registry: canonical

Reviewer 2: The point of the test case is to put in live data, so we should replace `{cborData}` with the actual encoded value

```

Multiple entities with metadata (Schema ID 2):

```

> Input: 0xdddddddd{cborData}00850280218021802180218021802180218021

Reviewer 2: Replace `{cborData}` with actual encoded value from your example
